### PR TITLE
Integrating API to UI:

### PR DESCRIPTION
**Pull request by dungvt9691**
- Sign in
- Sign up
- Sign out

### CODE REVIEW

File: lib/vue_phoenix/models/user.ex [hunk 1]
```diff
@@ -32,11 +32,16 @@ defmodule VuePhoenix.User do
...
-        token = Authenticator.generate_token(user)
-        Repo.insert(Ecto.build_assoc(user, :tokens, %{code: token}))
-      err -> err
+    case Repo.get_by(User, %{email: email}) do
+      nil -> {:error, %{email: "Your email address is not registered"}}
+      user ->
+        case Encryption.validate_password(user, password) do
+          {:ok, user} ->
+            token = Authenticator.generate_token(user)
+            Repo.insert(Ecto.build_assoc(user, :tokens, %{code: token}))
```
Comment (kimquy) [inline lib/vue_phoenix/models/user.ex]: Not require but this is the common suggested style 

```elixir
user
|> Ecto.build_assoc(:tokens, %{code: token})
|> Repo.insert
```

**Comment by kimquy** [comment]
@dungvt9691 Please add the tests for those API also

File: lib/vue_phoenix/models/user.ex (hunk 1 above)
Comment (dungvt9691) [reply lib/vue_phoenix/models/user.ex]: Yeah, I forgot it. I'm still familiar with the syntax of rails

**Comment by dungvt9691** [comment]
@kimquy I added test

File: lib/vue_phoenix/models/user.ex [hunk 2]
```diff
@@ -26,27 +26,43 @@ defmodule VuePhoenix.User do
     |> cast(attrs, [:email, :password])
     |> validate_required([:email, :password])
     |> unique_constraint(:email)
+    |> validate_format(:email, ~r/(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)/)
```
Comment (kimquy) [inline lib/vue_phoenix/models/user.ex]: can refactor this `~r/(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)/`  into `module attribute`

File: lib/vue_phoenix/models/user.ex [hunk 3]
```diff
@@ -26,27 +26,43 @@ defmodule VuePhoenix.User do
...
     |> encrypt_password
   end
 
-  def sign_in(email, password) do
-    case Encryption.validate_password(Repo.get_by(User, email: email), password) do
-      {:ok, user} ->
-        token = Authenticator.generate_token(user)
-        Repo.insert(Ecto.build_assoc(user, :tokens, %{code: token}))
-      err -> err
+  def sign_in(%{email: email, password: password}) do
```
Comment (kimquy) [inline lib/vue_phoenix/models/user.ex]: Not ideal to have `sign_in` or `sign_out` at the model layer. Can use service module for this or apply DDD. 

For example:

```elixir
defmodule Authentication
  def sign_in()
    ...
  end

  def sign_out
    ...
  end
end
```

Comment (kimquy) [reply lib/vue_phoenix/models/user.ex]: Then can use `Authentication.sign_in ` -> more clean

File: test/support/factory.ex [hunk 4]
```diff
@@ -0,0 +1,18 @@
+defmodule VuePhoenix.Factory do
+  use ExMachina.Ecto, repo: VuePhoenix.Repo
+
+  alias VuePhoenix.{User, Token}
+
+  def user_factory do
+    %User{
+      email: "dungvt9691@gmail.com",
```
Comment (kimquy) [inline test/support/factory.ex]: Can use this `https://github.com/igas/faker` to generate test data

File: test/vue_phoenix/models/token_test.exs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+defmodule VuePhoenix.TokenTest do
+  use VuePhoenix.DataCase
+
+  alias VuePhoenix.{User, Token, Repo}
+
+  describe "changeset" do
+    @valid_attrs %{code: "code", user_id: 1}
+    @invalid_attrs %{code: ""}
+    @user_attrs %{email: "dungvt9691@gmail.com", password: "password", password_confirmation: "password"}
```
Comment (kimquy) [inline test/vue_phoenix/models/token_test.exs]: Not ideal to use `module attribute`

can simply do simple assignment

`user_attrs = %{email: "dungvt9691@gmail.com", password: "password", password_confirmation: "password"}`

Comment (kimquy) [reply test/vue_phoenix/models/token_test.exs]: If you want to create share test environemtn then can use `setup` block

For example:

```elixir
setup do
  user_attrs = %{email: "dungvt9691@gmail.com", password: "password", password_confirmation: "password"}
  [user_attrs: user_attrs]
end
```

Comment (kimquy) [reply test/vue_phoenix/models/token_test.exs]: Then can be easily reuse in all blocks

**Comment by dungvt9691** [comment]
@kimquy, I update test follow your feedback and also applied DDD. 

But I have a small question: Why we shouldn't use the module attribute? I see from Phoenix docs to some tutorials, they usually use the module attribute instead of using setup.

**Pull request merged**